sky/utils/vpn_utils.py

```diff
@@ -0,0 +1,309 @@
+        """Create a VPN configuration from the backend configuration."""
+        vpn_configs = {}
+        for vpn_name, vpn_config in config.items():
+            assert vpn_name in ['tailscale'
+                               ], (f'Unknown VPN configuration: {vpn_name}')
+            if vpn_name == 'tailscale':
+                vpn_configs[vpn_name] = \
+                    TailscaleVPN.from_backend_config(vpn_config)
+        return VPNConfig(vpn_configs)
+
+    def get_setup_commands(self, hostname: str) -> List[str]:
+        """Get the command to setup the VPN on VM instances."""
+        return [
+            vpn_config.get_setup_command(hostname)
+            for vpn_config in self._vpn_configs.values()
+        ]
+
+    def get_setup_env_vars(self) -> Dict[str, str]:
+        """Get the environment variables to setup instances with the VPN."""
+        env_vars = {}
+        for vpn_config in self._vpn_configs.values():
+            env_vars.update(vpn_config.get_setup_env_vars())
+        return env_vars
+
+    def get_private_ip(self, hostname: str) -> Dict[str, str]:
+        """Get the private IP address from the hostname."""
+        private_ips = {}
+        for vpn_name, vpn_config in self._vpn_configs.items():
+            private_ip = vpn_config.get_private_ip(hostname)
+            private_ips[vpn_name] = private_ip
+        return private_ips
+
+    def get_private_ip_vpn(self, hostname: str, vpn_name: str) -> str:
+        """Get the private IP address from the hostname for a specific VPN."""
+        return self._vpn_configs[vpn_name].get_private_ip(hostname)
+
+    def remove_host(self, hostname: str) -> None:
+        """Remove a host from the VPN."""
+        for vpn_config in self._vpn_configs.values():
+            vpn_config.remove_host(hostname)
+
+    def to_yaml_config(self) -> Dict[str, Any]:
+        """Get the VPN configuration in YAML format."""
+        vpn_config_dict = {}
+        for vpn_name, vpn_config in self._vpn_configs.items():
+            vpn_yaml = vpn_config.to_yaml_config()
+            if vpn_yaml:
+                vpn_config_dict[vpn_name] = vpn_yaml
+            else:
+                vpn_config_dict[vpn_name] = True
+        return vpn_config_dict
+
+    def to_backend_config(self) -> Dict[str, Any]:
+        """Generate the VPN configuration for the backend."""
+        vpn_config_dict = {}
+        for vpn_name, vpn_config in self._vpn_configs.items():
+            vpn_config_dict[vpn_name] = vpn_config.to_backend_config()
+        return vpn_config_dict
+
+    def __getitem__(self, key: str) -> VPN:
+        return self._vpn_configs[key]
+
+    def __setitem__(self, key: str, value: VPN) -> None:
+        self._vpn_configs[key] = value
+
+
+def rewrite_cluster_info_by_vpn(
+    cluster_info: common.ClusterInfo,
+    vpn_config: VPNConfig,
+) -> common.ClusterInfo:
+    for (instance_id, instance_list) in cluster_info.instances.items():
+        # TODO(yi): test if this works on TPU VM.
```

Comment:
Lets raise an error for TPU VM + vpn config first. Also pls submit an issue for that ;)